extensions/amp-audio/1.0/component.js

```diff
@@ -0,0 +1,195 @@
+ * limitations under the License.
+ */
+
+import * as Preact from '#preact';
+import {listen} from '../../../src/event-helper';
+import {setMediaSession} from '../../../src/mediasession-helper';
+import {forwardRef} from '#preact/compat';
+import {triggerAnalyticsEvent} from '../../../src/analytics';
+
+const {useCallback, useEffect, useImperativeHandle, useMemo, useRef} = Preact;
+
+/**
+ * @param {!AudioDef.Props} props
+ * @param {{current: (!AudioDef.AudioApi|null)}} ref
+ * @return {PreactDef.Renderable}
+ */
+export function AudioWithRef(props, ref) {
+  const audioRef = useRef();
+
+  const {
+    album,
+    artist,
+    artwork,
+    autoplay,
+    children,
+    controlsList,
+    isStoryDescendant_,
+    loop,
+    muted,
+    preload,
+    propagateAttributes,
+    src,
+    title,
+    toggleFallback,
+    validateMediaMetadata,
+    ...rest
+  } = props;
+
+  /** @public {boolean} */
+  const isPlaying = useRef(false);
+
+  /**
+   * Checks if the function is allowed to be called
+   * @return {boolean}
+   */
+  const isInvocationValid_ = useCallback(() => {
+    if (!audioRef) {
+      return false;
+    }
+
+    if (isStoryDescendant_ && isStoryDescendant_()) {
+      console /*OK*/
+        .warn(
+          '<amp-story> elements do not support actions on <amp-audio> elements'
+        );
+      return false;
+    }
+    return true;
+  }, [isStoryDescendant_]);
+
+  /**
+   * Prepares Media Metadata
+   */
+  const metaData = useMemo(() => {
+    return {
+      title,
+      artist,
+      album,
+      artwork: [{src: artwork}],
+    };
+  }, [title, artist, album, artwork]);
+
+  /**
+   * Plays audio callback
+   */
+  const playCallback = useCallback(() => {
+    triggerAnalyticsEvent(audioRef.current, 'audio-play');
+
+    if (!isInvocationValid_()) {
+      return;
+    }
+    audioRef.current.play();
+    isPlaying.current = true;
+  }, [isPlaying, isInvocationValid_]);
+
+  /**
+   * Pauses audio callback
+   */
+  const pauseCallback = useCallback(() => {
+    triggerAnalyticsEvent(audioRef.current, 'audio-pause');
+
+    if (!isInvocationValid_()) {
+      return;
+    }
+    audioRef.current.pause();
+    isPlaying.current = false;
+  }, [isPlaying, isInvocationValid_]);
+
+  /**
+   * Updates media session for current window/tab
+   */
+  const audioPlaying = useCallback(() => {
+    const win = audioRef.current?.ownerDocument?.defaultView;
+    const element = audioRef.current;
+
+    if (validateMediaMetadata) {
+      validateMediaMetadata(element, metaData);
+    }
+
+    setMediaSession(win, metaData, playCallback, pauseCallback);
+  }, [metaData, validateMediaMetadata, playCallback, pauseCallback]);
+
+  useEffect(() => {
+    if (!audioRef.current.play) {
+      if (toggleFallback) {
+        toggleFallback(true);
+      }
+      return;
+    }
+
+    const unlistenPlaying = listen(audioRef.current, 'playing', () =>
+      audioPlaying()
+    );
```

Comment:
We don't want to use the `listen` helper because it's the imperative way of attaching event handlers. Can we instead use the [`onPlaying`](https://reactjs.org/docs/events.html#media-events) prop?

Reply:
Absolutely. Let me implement `onPlaying` instead of `listen`.